auth/authentication.go

```diff
@@ -17,119 +16,142 @@ type Database interface {
 	GetApplicationByToken(token string) (*model.Application, error)
 	GetClientByToken(token string) (*model.Client, error)
 	GetPluginConfByToken(token string) (*model.PluginConf, error)
+
 	GetUserByName(name string) (*model.User, error)
 	GetUserByID(id uint) (*model.User, error)
+	CreateUser(user *model.User) error
+}
+
+// AuthenticationProvider provides authentication methods
+type AuthenticationProvider interface {
+	Name() string
+	Authenticate(req *http.Request) (user *model.User, err error)
 }
 
 // Auth is the provider for authentication middleware
 type Auth struct {
-	DB Database
+	authenticators []AuthenticationProvider
+	DB             Database
 }
 
-type authenticate func(tokenID string, user *model.User) (authenticated bool, success bool, userId uint, err error)
+// RegisterAuthenticationProvider registers a new authentication provider
+// the provider is used in the order it i registered
+func (a *Auth) RegisterAuthenticationProvider(auth AuthenticationProvider) {
+	a.authenticators = append(a.authenticators, auth)
+}
 
-// RequireAdmin returns a gin middleware which requires a client token or basic authentication header to be supplied
+func (a *Auth) getUserInfoFromAuth(ctx *gin.Context) (user *model.User, authenticator string, err error) {
+	designatedAuthenticator := ctx.GetHeader("X-Gotify-Authenticator")
+
+	for _, authProvider := range a.authenticators {
+		providerName := authProvider.Name()
+		if designatedAuthenticator != "" && providerName != designatedAuthenticator {
+			continue
+		}
+		if user, err := authProvider.Authenticate(ctx.Request); user != nil {
+			return user, providerName, nil
+		} else if err != nil {
+			return nil, providerName, err
+		}
+	}
+	return nil, "", NoAuthProviderError{}
+}
+
+// ObtainAuthentication attempts to get user authentication from either client token or an authentication provider
+func (a *Auth) ObtainAuthentication(ctx *gin.Context) (user *model.User, err error) {
+	// 1. check for existing client token
+	if token := tokenFromQueryOrHeader(ctx); token != "" {
+		client, _ := a.DB.GetClientByToken(token)
+		if client == nil {
+			return nil, TokenRequiredError{"client"}
+		}
+		user, _ := a.DB.GetUserByID(client.UserID)
+		if user == nil {
+			return nil, TokenRequiredError{"client"}
+		}
+		return user, nil
+	}
+
+	// 2. try user authentication
+	user, authenticator, err := a.getUserInfoFromAuth(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if user.ID == 0 {
```

Comment:
If the user does exist, we need to verify that `user.Authenticator == authenticator`.